### PR TITLE
Feature: Carbon DeFi

### PR DESCRIPTION
**Pull request by tiagofilipenunes**
# Add CarbonDeFi

## Useful links

Website: https://carbondefi.xyz
App: https://app.carbondefi.xyz

Socials: 
- Twitter: https://twitter.com/carbondefixyz
- Discord: https://discord.gg/bancor
- Telegram: https://t.me/CarbonDeFixyz
- Medium: https://blog.carbondefi.xyz/
- Youtube: https://www.youtube.com/c/BancorProtocol

Carbon DeFi documentation: https://docs.carbondefi.xyz/

CarbonSDK (not used here but can prove useful in QA): https://github.com/bancorprotocol/carbon-sdk

## Summary of protocol

CarbonDeFi is a decentralised protocol live on Ethereum mainnet with similar deployments on Base, Mantle, Canto, Arbitrum, and Fantom. CarbonDeFi supports any ERC-20 token (sans fee-on-transfer and rebase tokens, etc). Carbon DeFi allows any user to:
- Create on-chain one-sided BUY/SELL range and limit orders for a pair of ERC-20 tokens ("disposable strategy")
- Create two on-chain connected BUY and SELL range and limit orders ("recurring strategy") for a pair of ERC-20 tokens. Each order in the strategy adds to the liquidity of the other when selling tokens at the range the user selects.
- Trade against any on-chain orders to obtain the best price execution (calculated by the Carbon DeFi SDK).
- Create overlapping strategies (concentrated liquidity-style)

Each Carbon DeFi position is represented by an NFT and called a Strategy, which consists of two orders. Each order contains 4 parameters (y, z, A, B) which represent the order's bonding curve geometric parameters, and can be used to deduce the order's balance, lowest rate, highest rate, and marginal rate. The parameter y represents the balance of each order.

All the tokens from the CarbonDeFi strategies are in the CarbonController contract (0xC537e898CD774e2dCBa3B14Ea6f34C93d5eA45e1), which is the main point of entrance for CarbonDeFi's smart contracts.

Since there are now forks of CarbonDeFi on other chains, the adapter was written to be easily expandable to new chains.

Not implemented:
- getProtocolTokens
- getTotalValueLocked
- getProtocolTokenToUnderlyingTokenRate
- getApy
- getApr
- fetchProtocolTokenMetadata (part of the acceptance criteria) - Each position/strategy is represented by an NFT, so this function wasn't implemented (similarly to the Uniswap v3 Pool adapter)
- fetchUnderlyingTokensMetadata
- getUnderlyingTokenBalances
- getUnderlyingTokenConversionRate
- getProtocolTokens

## Methods for calculating information to facilitate QA
- Positions
  1. Get strategyIds from the Voucher NFT Contract
  2. For each strategy represented by strategyId, get the balance of each order from the Carbon Controller contract
- getWithdrawals/getDeposits
  1. Get StrategyCreated, StrategyUpdated, and StrategyDeleted events. 
  2. Process the events. Each event contains the order's new y, or the order's y before deletion for the case of StrategyDeleted. This means that in order to assess differences in balance, the events must be compared over time. For strategyCreated, the new y represents the deposited balance for each order. For strategyDeleted, the y in the event is the balance before deleted the strategy and represents the withdrawn balance for each order. The StrategyUpdated events contain a parameter reason which can be 0 (edit by a user meaning a potential deposit/withdrawal) or 1 (strategy updated as a consequence of a trade being routed through it). Only the StrategyUpdated with reason===0 are relevant for user deposits and withdrawals, but the all events must still be considered to measure the balance differences. Another approach could have been to use the Transfer event from the user to the carbonController contract, being careful that a user can create two different strategies for the same pair/involved token.
  3. For a withdrawal, count the strategyDeleted y for each order and the withdrawals obtained from the StrategyCreated and StrategyUpdated events. For a deposit, count the strategyCreated y for each order and the deposits obtained from the StrategyCreated and the StrategyUpdated events.

Thank you in advance for the time spend reviewing and for the thorough documentation, making it easy for external contributors to add code to the repo!

### CODE REVIEW

File: src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts
Comment (bergarces) [inline src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts:56]: Since the methods of SimplePoolAdapter are not being used, it'd be better to `implements IProtocolAdapter` instead of `extends SimplePoolAdapter`.

That way you should also be able to remove the protected abstract methods from `SimplePoolAdapter`, as they are not being used.

**Comment by bergarces** [comment]
Hello @tiagofilipenunes . Thanks for the very detailed and documented PR. The code looks good and it's almost ready to be merged.

If you could edit the PR so that we maintainers have permission to make changes, I'm happy to update it myself, but otherwise this is what I'd like to do before merging:
- Implement `IProtocolAdapter` instead of extending `SimplePoolAdapter`.
- Completely remove `fetchProtocolTokenMetadata`, `fetchUnderlyingTokensMetadata`, `getUnderlyingTokenBalances` and `getUnderlyingTokenConversionRate` instead of throwing, since they are not being used.
- Remove the `update me` comments on top of methods, as they were supposed to only be a guide (we are going to remove those comments from the template soon).

Other than that, code and tests look good to go.

File: src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts
Comment (bergarces) [inline src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts:338]: It looks like this is the only item of the enum we are using.

Could we instead of declaring an enum, create a constant such as:
`const StrategyUpdateReasonEdit = 0n`

That way no conversion is needed either.

**Comment by tiagofilipenunes** [comment]
Thank you for reviewing this PR so quickly @bergarces !

I've just addressed your requested changes, please check that everything is okay. I can't change the permissions for this PR so maintainers can edit since it is an organization-wide setting - as the fork is in the bancorprotocol org - but let me know if you still need it as I can ask the right person to change this.

**Review by bergarces — APPROVED** [review]

**Pull request merged**
